README.md
# vscode-author-generator

A tool to generate Custom Author Info at the top of your file.

![gif](https://raw.githubusercontent.com/excaliburhan/vscode-author-generator/master/example.gif)

## Installation

Press `F1` in VSCode, type `ext install` and then look for `vscode-author-generator`.

## Features

- Support multiple templates, `js`, `css`, `html/htm` file types are supported for now.
- Support custom infomation, `author name`, `email`, `date` info is supported for now.
- Add `cmd+u` keybindings for quick update.

## Usage

### Before Use

- `Preferences` -> `Settings` to add your own settings
- `author-generator.author` author name, will replace `[author]` in your `*.tpl` file
- `author-generator.email`. author email, will replace `[email]` in your `*.tpl` file
- `author-generator.dateFormat` will change your date format(default format is `YYYY-MM-DD HH:mm:ss`)
- `author-generator.updateOnSave` set `true` to trigger updateInfo on save, only `*.tpl` template files exsit will trigger

### Commands

- Generate Author Info - generate your author info right now.

### Manage Custom Templates

#### 1. Open extension directory and find `vscode-author-generator`

- Windows: %USERPROFILE%\.vscode\extensions\{edwardhjp.vscode-author-generator-version}
- Mac: ~/.vscode/extensions/{edwardhjp.vscode-author-generator-version}
- Linux: ~/.vscode/extensions/{edwardhjp.vscode-author-generator-version}

#### 2. Edit or add template file

- `vscode-author-generator` use file suffix as template name. eg: `js.tpl` is the template for `*.js` files. if the specific template file doesn't exsit, `default.tpl` will be used.
- if you want to add another template, just make sure the template filename is correct. eg: add a file named `py.tpl` in `templates` directory will make `*.py` files add this template info.
- notice: `[author]`, `[email]` will automatically replaced with infomation in your VSCode Configration.


CHANGELOG.md
### V0.2.2

- [Fix] fix vscode document is not found

### V0.2.1

- [Feature] updateOnSave will not trigger if there is no `@modify date` string

### V0.2.0

- [Feature] add `author-generator.updateOnSave`, you can trigger updateInfo on save automaticly

### V0.0.10

- [Feature] add `author-generator.dateFormat` to custom the format of date

### V0.0.7

- [Feature] add `cmd+u` keybings for quick update

### V0.0.5

- [Fix] fix time format

### V0.0.4

- [Feature] add `Update Auth Info` contribute command

### V0.0.3

- [Feature] add `example.gif` to README.md

### V0.0.2

- [Fix] remove whitespaces in template files

### V0.0.1

- Initial release


extension.js
/**
 * @author xiaoping
 * @email edwardhjp@gmail.com
 * @create date 2017-02-13 11:51:32
 * @modify date 2020-08-07 13:03:51
 * @desc [extension file]
 */

let vscode = require('vscode');
let generator = require('./generator.js');

function activate(context) {
  let initAuthor = vscode.commands.registerCommand('extension.generateAuthorInfo', () => {
    generator.initInfo();
  });

  let updateAuthor = vscode.commands.registerCommand('extension.updateAuthorInfo', () => {
    generator.updateInfo();
  });

  context.subscriptions.push(initAuthor);
  context.subscriptions.push(updateAuthor);
}

function deactivate() {}

function initUpdateSave() {
  vscode.workspace.onWillSaveTextDocument(() => {
    generator.updateOnSave();
  });
}

initUpdateSave();

exports.activate = activate;
exports.deactivate = deactivate;


generator.js
/**
 * @author xiaoping
 * @email edwardhjp@gmail.com
 * @create date 2017-02-10 13:10:00
 * @modify date 2020-08-16 20:35:38
 * @desc [generator file]
 */

let vscode = require('vscode');
let path = require('path');
let fs = require('fs');
let moment = require('moment');

module.exports = {
  initInfo() {
    let editor = vscode.window.activeTextEditor;
    editor.edit((builder) => {
      try {
        let document = editor.document || editor._documentData._document;
        let tplText = this.getTplText(document);
        builder.insert(new vscode.Position(0, 0), tplText);
      } catch (error) {
        vscode.window.showErrorMessage(error.message);
      }
    });
  },
  updateInfo() {
    let editor = vscode.window.activeTextEditor;
    let range = new vscode.Range(new vscode.Position(0, 0), new vscode.Position(21, 0));
    let text = editor.document.getText(range);
    let modifyIndex = text.indexOf('@modify date ');
    let modifiedPos = editor.document.positionAt(modifyIndex);
    // if there is no @modify date, just return
    if (modifyIndex === -1) {
      return;
    }
    let modifiedLine = modifiedPos.line;
    let line = editor.document.lineAt(modifiedLine);
    let lineText = line.text;
    let lineTextArr = lineText.split('@modify date ');
    let replaceText = '';
    for (let i = 0; i < lineTextArr.length - 1; i++) {
      replaceText += lineTextArr[i];
    }
    replaceText += '@modify date ' + this.getDate();
    editor.edit((builder) => {
      try {
        builder.replace(line.range, replaceText);
      } catch (error) {
        vscode.window.showErrorMessage(error.message);
      }
    });
  },
  updateOnSave() {
    let config = this.getConfig();
    if (config.updateOnSave) {
      try {
        let editor = vscode.window.activeTextEditor;
        let document = editor.document || editor._documentData._document;
        let fileType = this.getFileType(document);
        let hasTpl = this.hasTplPath(fileType);
        if (hasTpl) {
          this.updateInfo();
        }
      } catch (error) {
        vscode.window.showErrorMessage(error.message);
      }
    }
  },
  getFileType(document) {
    let fileInfo = document.fileName.split('.');
    return fileInfo.length > 1 ? fileInfo.pop() : 'default';
  },
  getTplPath(type) {
    type = type.toLowerCase();
    let extDir = vscode.extensions.getExtension('edwardhjp.vscode-author-generator').extensionPath;
    let extPath = path.join(extDir, 'templates', `${type}.tpl`);
    if (fs.existsSync(extPath)) {
      return extPath;
    } else {
      return path.join(extDir, 'templates', 'default.tpl');
    }
  },
  hasTplPath(type) {
    type = type.toLowerCase();
    let extDir = vscode.extensions.getExtension('edwardhjp.vscode-author-generator').extensionPath;
    let extPath = path.join(extDir, 'templates', `${type}.tpl`);
    if (fs.existsSync(extPath)) {
      return true;
    }
    return false;
  },
  getTplText(document) {
    let text = '';
    let config = this.getConfig();
    let type = this.getFileType(document);
    let tplPath = this.getTplPath(type);
    try {
      text = fs.readFileSync(tplPath, 'utf-8');
      text = text
        .replace(/\[author\]/, config.author)
        .replace(/\[email\]/, config.email)
        .replace(/\[date\]/g, config.date);
    } catch (error) {
      vscode.window.showErrorMessage(error.message);
    }
    return text;
  },
  getConfig() {
    let config = vscode.workspace.getConfiguration('author-generator');
    config = {
      author: config.get('author'),
      email: config.get('email'),
      date: this.getDate(),
      updateOnSave: config.get('updateOnSave')
    };
    return config;
  },
  getDate() {
    let config = vscode.workspace.getConfiguration('author-generator');
    let dateFormat = config.get('dateFormat') || 'YYYY-MM-DD HH:mm:ss';
    return moment().format(dateFormat);
  }
};


rollup.config.js
import path from 'path';
import rollupTypescript from 'rollup-plugin-typescript2';
import babel from '@rollup/plugin-babel';
import resolve from '@rollup/plugin-node-resolve';
import commonjs from '@rollup/plugin-commonjs';
import { DEFAULT_EXTENSIONS } from '@babel/core';
import pkg from './package.json';

const paths = {
  input: path.join(__dirname, '/src/extension.ts'),
  output: path.join(__dirname, '/dist')
};

const rollupConfig = {
  input: paths.input,
  output: [
    {
      file: path.join(paths.output, 'extension.js'),
      format: 'cjs',
      name: pkg.name,
      exports: 'named'
    }
  ],
  // external: ['moment'],
  plugins: [
    commonjs(),
    resolve(),
    rollupTypescript(),
    babel({
      babelHelpers: 'bundled',
      exclude: 'node_modules/**',
      extensions: [...DEFAULT_EXTENSIONS, '.ts']
    })
  ]
};

export default rollupConfig;


src/extension.ts
/**
 * @author xiaoping
 * @email edwardhjp@gmail.com
 * @create date 2017-02-13 11:51:32
 * @modify date 2021-03-10 16:02:03
 * @desc [extension file]
 */

import * as vscode from 'vscode';
import generator from './generator';

function activate(context) {
  let initAuthor = vscode.commands.registerCommand('extension.generateAuthorInfo', () => {
    generator.initInfo();
  });

  let updateAuthor = vscode.commands.registerCommand('extension.updateAuthorInfo', () => {
    generator.updateInfo();
  });

  context.subscriptions.push(initAuthor);
  context.subscriptions.push(updateAuthor);
}

function deactivate() {}

function initUpdateSave() {
  (vscode.workspace as any).onWillSaveTextDocument(() => {
    generator.updateOnSave();
  });
}

initUpdateSave();

exports.activate = activate;
exports.deactivate = deactivate;


src/generator.ts
/**
 * @author xiaoping
 * @email edwardhjp@gmail.com
 * @create date 2017-02-10 13:10:00
 * @modify date 2021-03-10 16:02:24
 * @desc [generator file]
 */

let path = require('path');
let fs = require('fs');
import * as vscode from 'vscode';
import moment from 'moment';

export default {
  initInfo() {
    let editor = vscode.window.activeTextEditor;
    editor.edit((builder) => {
      try {
        let document = editor.document || (editor as any)._documentData._document;
        let tplText = this.getTplText(document);
        builder.insert(new vscode.Position(0, 0), tplText);
      } catch (error) {
        vscode.window.showErrorMessage(error.message);
      }
    });
  },
  updateInfo() {
    let editor = vscode.window.activeTextEditor;
    let range = new vscode.Range(new vscode.Position(0, 0), new vscode.Position(21, 0));
    let text = editor.document.getText(range);
    let modifyIndex = text.indexOf('@modify date ');
    let modifiedPos = editor.document.positionAt(modifyIndex);
    // if there is no @modify date, just return
    if (modifyIndex === -1) {
      return;
    }
    let modifiedLine = modifiedPos.line;
    let line = editor.document.lineAt(modifiedLine);
    let lineText = line.text;
    let lineTextArr = lineText.split('@modify date ');
    let replaceText = '';
    for (let i = 0; i < lineTextArr.length - 1; i++) {
      replaceText += lineTextArr[i];
    }
    replaceText += '@modify date ' + this.getDate();
    editor.edit((builder) => {
      try {
        builder.replace(line.range, replaceText);
      } catch (error) {
        vscode.window.showErrorMessage(error.message);
      }
    });
  },
  updateOnSave() {
    let config = this.getConfig();
    if (config.updateOnSave) {
      try {
        let editor = vscode.window.activeTextEditor;
        let document = editor.document || (editor as any)._documentData._document;
        let fileType = this.getFileType(document);
        let hasTpl = this.hasTplPath(fileType);
        if (hasTpl) {
          this.updateInfo();
        }
      } catch (error) {
        vscode.window.showErrorMessage(error.message);
      }
    }
  },
  getFileType(document) {
    let fileInfo = document.fileName.split('.');
    return fileInfo.length > 1 ? fileInfo.pop() : 'default';
  },
  getTplPath(type) {
    type = type.toLowerCase();
    let extDir = vscode.extensions.getExtension('edwardhjp.vscode-author-generator').extensionPath;
    let extPath = path.join(extDir, 'templates', `${type}.tpl`);
    if (fs.existsSync(extPath)) {
      return extPath;
    } else {
      return path.join(extDir, 'templates', 'default.tpl');
    }
  },
  hasTplPath(type) {
    type = type.toLowerCase();
    let extDir = vscode.extensions.getExtension('edwardhjp.vscode-author-generator').extensionPath;
    let extPath = path.join(extDir, 'templates', `${type}.tpl`);
    if (fs.existsSync(extPath)) {
      return true;
    }
    return false;
  },
  getTplText(document) {
    let text = '';
    let config = this.getConfig();
    let type = this.getFileType(document);
    let tplPath = this.getTplPath(type);
    try {
      text = fs.readFileSync(tplPath, 'utf-8');
      text = text
        .replace(/\[author\]/, config.author)
        .replace(/\[email\]/, config.email)
        .replace(/\[date\]/g, config.date);
    } catch (error) {
      vscode.window.showErrorMessage(error.message);
    }
    return text;
  },
  getConfig() {
    let config: any = vscode.workspace.getConfiguration('author-generator');
    config = {
      author: config.get('author'),
      email: config.get('email'),
      date: this.getDate(),
      updateOnSave: config.get('updateOnSave')
    };
    return config;
  },
  getDate() {
    let config: any = vscode.workspace.getConfiguration('author-generator');
    let dateFormat = config.get('dateFormat') || 'YYYY-MM-DD HH:mm:ss';
    return moment().format(dateFormat);
  }
};
